_source/_posts/2022-01-13-spring-cloud-contract.md

```diff
@@ -0,0 +1,893 @@
+
+Unzip the project and open it in the IDE or editor of your choice.
+
+Create a `Hat.java` class that will model the hat returned from the producer.
+
+`src/main/java/com/example/consumer/Hat.java`
+
+```java
+package com.example.consumer;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor // needed for Jackson serialization
+public class Hat {
+    private Long id;
+    private String name;
+    private Long size;
+    private String color;
+}
+```
+
+Update the `ConsumerApplication.java` class.
+
+`src/main/java/com/example/consumer/ConsumerApplication.java`
+
+```java
+package com.example.consumer;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.client.RestTemplate;
+
+@SpringBootApplication
+public class ConsumerApplication {
+
+    public static void main(String[] args) {
+        SpringApplication.run(ConsumerApplication.class, args);
+    }
+
+    @RestController
+    public class HatConsumerResource {
+
+        @Value( "${producer.port}" )
+        private Integer producerPort;
+
+        private final RestTemplate restTemplate;
+
+        HatConsumerResource(RestTemplateBuilder restTemplateBuilder) {
+            this.restTemplate = restTemplateBuilder.build();
+        }
+
+        @RequestMapping("/wearhat/{hatId}")
+        String getMessage(@PathVariable("hatId") Long personId, @RequestHeader("authorization") String authorizationHeader) {
+
+            HttpHeaders headers = new HttpHeaders();
+            headers.set("Authorization", authorizationHeader);
+            HttpEntity<Void> requestEntity = new HttpEntity<>(headers);
+
+            ResponseEntity<Hat> response = restTemplate.exchange("http://localhost:"+producerPort+"/hat/{hatId}", HttpMethod.GET, requestEntity, Hat.class, personId);
+            Hat hat = response.getBody();
+
+            return "Enjoy your new " + hat.getName();
+        }
+
+    }
+}
+```
+
+Add a line to the `application.properties` file to change the application port from the default, 8080, to 8081 so that it does not conflict with the producer application. You're also going to specify the producer port in the properties file so that the integration tests can use a different port for the mocked service (you'll see why in a bit).
```

Comment:
```suggestion
Add a line to the `application.properties` file to change the application port from the default, 8080, to 8081 so that it does not conflict with the producer application. You're also going to specify the producer port in the properties file so that the integration tests can use a different port for the mocked service. (You'll see why in a bit.)
```